src/table.rs

```diff
@@ -65,31 +70,146 @@ pub struct Table {
+            id,
+            checksum: Bytes::new(),
+            estimated_size: 0,
+            index: TableIndex::default(),
+            index_start: 0,
+            index_len: 0,
+            opt,
+        };
+        inner.init_biggest_and_smallest()?;
+        // TODO: verify checksum
+        Ok(inner)
     }
 
     /// Open an existing SST from data in memory
-    pub fn open_in_memory(_data: Bytes, _id: u64, _opt: Options) -> Result<TableInner> {
-        unimplemented!()
+    fn open_in_memory(data: Bytes, id: u64, opt: Options) -> Result<TableInner> {
+        let table_size = data.len();
+        let mut inner = TableInner {
+            file: MmapFile::Memory { data },
+            opt,
+            table_size,
+            id,
+            smallest: Bytes::new(),
+            biggest: Bytes::new(),
+            checksum: Bytes::new(),
+            estimated_size: 0,
+            index: TableIndex::default(),
+            index_start: 0,
+            index_len: 0,
+        };
+        inner.init_biggest_and_smallest()?;
+        Ok(inner)
     }
 
     fn init_biggest_and_smallest(&mut self) -> Result<()> {
-        unimplemented!()
+        let ko = self.init_index()?;
+        self.smallest = Bytes::copy_from_slice(&ko.key);
+        let mut it = TableIterator::new(&self, ITERATOR_REVERSED | ITERATOR_NOCACHE);
+        it.rewind();
+        if !it.valid() {
+            return Err(Error::TableRead(format!(
+                "failed to initialize biggest for table {}",
+                self.filename()
+            )));
+        }
+        self.biggest = Bytes::copy_from_slice(it.key());
+        Ok(())
     }
 
     fn init_index(&mut self) -> Result<&BlockOffset> {
-        unimplemented!()
+        let mut read_pos = self.table_size;
+
+        // read checksum length from last 4 bytes
+        read_pos -= 4;
+        let mut buf = self.read(read_pos, 4)?;
+        let checksum_len = buf.get_u32() as usize;
+        if (checksum_len as i32) < 0 {
+            return Err(Error::TableRead(
+                "checksum length less than zero".to_string(),
+            ));
+        }
+
+        // read checksum
+        read_pos -= checksum_len;
+        let buf = self.read(read_pos, checksum_len)?;
+        let chksum = Checksum::decode(buf)?;
+
+        // read index size from footer
+        read_pos -= 4;
+        let mut buf = self.read(read_pos, 4)?;
+        self.index_len = buf.get_u32() as usize;
+
+        // read index
+        read_pos -= self.index_len;
+        self.index_start = read_pos;
+        let data = self.read(read_pos, self.index_len)?;
+        checksum::verify_checksum(&data, &chksum)?;
+
+        self.index = self.read_table_index()?;
+
+        // TODO: compression
+        self.estimated_size = self.table_size as u32;
+
+        // TODO: has bloom filter
+
+        Ok(&self.index.offsets[0])
     }
 
     fn key_splits(&mut self, _n: usize, _prefix: Bytes) -> Vec<String> {
+        // if n == 0 {
```

Comment:
why comment out?

Reply:
This function is not implemented, and will be implemented later.

Reply:
I've removed the implementation for now